deepspeed/runtime/fp16/onebit/zoadam.py

```diff
@@ -0,0 +1,352 @@
+                    torch.cuda.empty_cache()
+                    state['worker_error'] = torch.zeros(state['corrected_tensor_size'],
+                                                        device=p.device)
+                    state['server_error'] = torch.zeros(state['server_chunk_size'],
+                                                        device=p.device)
+                    # Accumulation of momentum, i.e., the u variable in the 0/1 Adam paper
+                    state['momentum_accumulator'] = torch.zeros_like(p.data)
+                    torch.cuda.empty_cache()
+                    # self.freeze_key = True
+                    if not self.initialize and torch.distributed.get_rank() == 0:
+                        print("Cupy Buffers Initialized Successfully.")
+
+                exp_avg, exp_avg_sq = state['exp_avg'], state['exp_avg_sq']
+                comm_buffer = state['momentum_accumulator']
+                beta1, beta2 = group['betas']
+
+                state['step'] += 1
+
+                if self.initialize:
+                    if self.freeze_key is False:
+                        if state['step'] % state['var_interval'] == 0:
+                            dist.all_reduce(grad)
+                            grad.mul_(1 / dist.get_world_size())
+                            exp_avg_sq.mul_(beta2).addcmul_(1 - beta2, grad, grad)
+                            exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+                        else:
+                            if self.size > 1:
+                                with torch.no_grad():
+                                    grad_onebit = self.comm_backend_handle.compressed_allreduce(
+                                        grad,
+                                        state['worker_error'],
+                                        state['server_error'],
+                                        self.deepspeed.local_rank)
+                                    exp_avg.mul_(beta1).add_(1 - beta1, grad_onebit)
+                    else:
+                        exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+                        state['lrs'] += group['lr']
+                    grad = None
+
+                if not self.initialize:
+                    if self.size > 1:
+                        comm_buffer.set_(
+                            self.comm_backend_handle.compressed_allreduce(
+                                comm_buffer,
+                                state['worker_error'],
+                                state['server_error'],
+                                self.deepspeed.local_rank))
+                    if 'exp_avg_mask' in group:
+                        if comm_buffer.device != group['exp_avg_mask'].device:
+                            group['exp_avg_mask'] = group['exp_avg_mask'].to(
+                                device=comm_buffer.device)
+                        comm_buffer.mul_(group['exp_avg_mask'])
```

Comment:
This mask need to be applied after every compressed_allreduce that is related to the gradient (because 1-bit compression cannot handle the exact zeros). So (1) Here you can add another indentation to only apply mask when self.size >1. (2) We also need to apply the mask after the two compressed_allreduce at line 225 and 265.